src/components/UserInputBlock.tsx

```diff
@@ -0,0 +1,84 @@
+import { EventControlledContent } from './EventControlledContent'
+import { Formik, Form, Field, ErrorMessage } from 'formik'
+import { useState, useCallback } from 'react'
+import { mathifyElement } from '../lib/math'
+import * as Yup from 'yup'
+
+const DEFAULT_BUTTON_TEXT = 'Submit'
+const DEFAULT_TEXTAREA_ROWS = 3
+
+interface UserInputBlockProps {
+  content: string
+  prompt: string
+  ack: string
+  waitForEvent?: string
+  fireEvent?: string
+  buttonText?: string
+}
+
+interface InputFormValues {
+  response: string
+}
+
+export const UserInputBlock = ({ content, prompt, ack, waitForEvent, fireEvent, buttonText }: UserInputBlockProps): JSX.Element => {
+  const [responseSubmitted, setResponseSubmitted] = useState(false)
+
+  const schema = Yup.object({
+    response: Yup.string().trim().required('Please provide valid input')
+  })
+
+  const handleSubmit = async (values: InputFormValues): Promise<void> => {
+    setResponseSubmitted(true)
+    if (fireEvent !== undefined) {
+      const submitEvent = new CustomEvent(fireEvent)
+      document.dispatchEvent(submitEvent)
+    }
+  }
+
+  const contentRefCallback = useCallback((node: HTMLDivElement | null): void => {
+    if (node != null) {
+      mathifyElement(node)
+    }
+  }, [])
+
+  const maybePrompt = responseSubmitted
+    ? null
+    : (
+        <div ref={contentRefCallback} dangerouslySetInnerHTML={{ __html: prompt }} />
+      )
+
+  const maybeInputForm = responseSubmitted
+    ? null
+    : (
+        <Formik
+          initialValues={{ response: '' }}
+          onSubmit={handleSubmit}
+          validationSchema={schema}
+        >
+          {({ isSubmitting }) => (
+            <Form>
+              <Field name="response" as="textarea" disabled={isSubmitting} rows={DEFAULT_TEXTAREA_ROWS} className="form-control my-3"/>
+              <ErrorMessage className="text-danger my-3" component="div" name="response" />
+              <button type="submit" disabled={isSubmitting} className="btn btn-outline-primary">{buttonText !== undefined ? buttonText : DEFAULT_BUTTON_TEXT}</button>
+            </Form>
+          )}
+        </Formik>
```

Comment:
Yeah, `formik` has a lot of magic that it does for you which is super useful and results in doing more with less code, but it may not make sense right away. I'd encourage you to go / read through [their tutorial](https://formik.org/docs/tutorial) just to get a better feel for it since we'll likely be using it in other components (e.g. problem sets) (cc: @P-Gill97 ).

To your question: `isSubmitting` is part of the data bag that `formik` provides (see [doc notes](https://formik.org/docs/api/formik#issubmitting-boolean)). That's being used here as a way to disable the button and input box while submission is happening. For this form there's no network / latency so it's probably impossible to see it, but the case this type of disabling is targeted to address is to not allow the user to accidentally submit a form while it's being processed (e.g. if we were doing a network request and the user clicked the submit button a 100 times for fun or something).

There is a bit of magical "mapping variables to the form fields" which is happening as well (e.g. the `name="response"` is how `formik` ties things to the `InputFormValues` interface I define in this file), but you'll get a better feel for that if you go through the tutorial.